core/src/main/java/org/apache/iceberg/Partitioning.java

```diff
@@ -228,11 +229,22 @@ public static StructType partitionType(Table table) {
           fieldMap.put(fieldId, field);
           NestedField structField = spec.partitionType().field(fieldId);
           structFields.add(structField);
+
+          if (Transforms.alwaysNull().equals(field.transform())) {
+            fieldToIndex.put(fieldId, structFields.size() - 1);
+          }
         } else {
           // verify the fields are compatible as they may conflict in v1 tables
           ValidationException.check(equivalentIgnoringNames(field, existingField),
               "Conflicting partition fields: ['%s', '%s']",
               field, existingField);
+
+          if (Transforms.alwaysNull().equals(existingField.transform()) &&
+                  !Transforms.alwaysNull().equals(field.transform())) {
+            fieldMap.put(fieldId, field);
+            NestedField structField = spec.partitionType().field(fieldId);
+            structFields.set(fieldToIndex.get(fieldId), structField);
```

Comment:
I think this may be easier (and we wouldn't need to build up the index map) to just do
```java
fieldMap.put(fieldId, field);
NestedField previousTransform  = spec.partitionType().field(fieldId);
voidTransformIndex = structFields.indexOf(existingField)
structFields.set(voidTransformIndex, previousTransform)
```

I recommend this because I think it more clearly expresses what we are trying to do here. When initially looked at the code It wasn't clear to me that the fieldToIndex was solely for looking up old void transforms for this edge case.

Reply:
```java
fieldMap.put(fieldId, field); 
NestedField previousTransform = spec.partitionType().field(fieldId); 
int voidTransformIndex = structFields.indexOf(existingField);  // here, we need a `NestedField` instance not a `PartitionField` instance, and it seems that we could not get the previous `PartitionField` because we are in new spec.
structFields.set(voidTransformIndex, previousTransform)```

Please see the above inline comments.

Reply:
ok got it, we need the "spec.partitionType()" of the spec the partition field was created with, and at this point we have lost that reference

Reply:
Yeah, that's right.